File: 2023/5/part2.py
from enum import Enum


class Maps(Enum):
    SEED_TO_SOIL = 1
    SOIL_TO_FERTILIZER = 2
    FERTILIZER_TO_WATER = 3
    WATER_TO_LIGHT = 4
    LIGHT_TO_TEMPERATURE = 5
    TEMPERATURE_TO_HUMIDITY = 6
    HUMIDITY_TO_LOCATION = 7


def resolve_map_range(map, input_range) -> int:
    resolved_range = []

    for row in map:
        # if map range starts before input_range
        if row[1] <= input_range[0] and input_range[0] < (row[1] + row[2]):
            #
            offset = input_range[0] - row[1]

            # if map range ends after input_range
            if (input_range[0] + input_range[1]) < (row[1] + row[2]):
                resolved_range.append((row[0] + offset, input_range[1]))
            # else map range ends before input_range
            else:
                resolved_range.append((row[0] + offset, row[2] - offset))

        # elif map range starts after input_range and before input_range ends
        elif input_range[0] < row[1] and row[1] < (input_range[0] + input_range[1]):
            #
            offset = row[1] - input_range[0]
            # if map range ends after input_range
            if (input_range[0] + input_range[1]) < (row[1] + row[2]):
                resolved_range.append((row[0], input_range[1] - offset))
            # else map range ends before input_range
            else:
                resolved_range.append((row[0], row[2]))

    # if unmapped
    if resolved_range == []:
        resolved_range.append(input_range)

    return resolved_range


min_location = 9223372036854775807  # 2**63-1 (hopefully will be lower)


def depth_firs_resolver(depth, range_to_resolve):
    global min_location
    next_category_ranges = resolve_map_range(maps[Maps(depth).name], range_to_resolve)
    if depth == Maps.HUMIDITY_TO_LOCATION.value:
        for location_range in next_category_ranges:
            if location_range[0] < min_location:
                min_location = location_range[0]
                print(f"new minimum: {min_location}")
    else:
        next_depth = depth + 1
        for next_category_range in next_category_ranges:
            depth_firs_resolver(next_depth, next_category_range)


seed_ranges = []
maps = {}
for enum in Maps:
    maps[enum.name] = []

with open("2023/5/input.txt", "r") as file:
    currently_reading_map = ""
    for i, line in enumerate(file):
        if line == "\n":
            continue
        if i == 0:
            seeds_str = line[6:-1].split()
            for j in range(0, len(seeds_str), 2):
                seed_ranges.append((int(seeds_str[j]), int(seeds_str[j + 1])))
            continue
        if line == "seed-to-soil map:\n":
            currently_reading_map = maps[Maps.SEED_TO_SOIL.name]
            continue
        if line == "soil-to-fertilizer map:\n":
            currently_reading_map = maps[Maps.SOIL_TO_FERTILIZER.name]
            continue
        if line == "fertilizer-to-water map:\n":
            currently_reading_map = maps[Maps.FERTILIZER_TO_WATER.name]
            continue
        if line == "water-to-light map:\n":
            currently_reading_map = maps[Maps.WATER_TO_LIGHT.name]
            continue
        if line == "light-to-temperature map:\n":
            currently_reading_map = maps[Maps.LIGHT_TO_TEMPERATURE.name]
            continue
        if line == "temperature-to-humidity map:\n":
            currently_reading_map = maps[Maps.TEMPERATURE_TO_HUMIDITY.name]
            continue
        if line == "humidity-to-location map:\n":
            currently_reading_map = maps[Maps.HUMIDITY_TO_LOCATION.name]
            continue

        row = []
        for number in line.split():
            row.append(int(number))
        currently_reading_map.append(row)

for seed_range in seed_ranges:
    depth_firs_resolver(1, seed_range)


print(min_location)

File: 2023/5/part2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <sstream>
#include <memory>

struct Range
{
    const uint start;
    const uint length;

    Range(const uint s, const uint l) : start(s), length(l){};
};

struct MapRange
{
    const uint destination_start;
    const uint source_start;
    const uint length;

    MapRange(const uint d, const uint s, const uint l) : destination_start(d), source_start(s), length(l){};
};

typedef std::shared_ptr<Range> Range_Ptr;
typedef std::vector<std::shared_ptr<Range>> Ranges;

typedef std::shared_ptr<MapRange> MapRange_Ptr;
typedef std::vector<std::shared_ptr<MapRange>> Map;
typedef std::vector<std::vector<std::shared_ptr<MapRange>>> Maps;

void readInputFile(const std::string &file_location, Ranges &seeds, Maps &maps);
void depthFirstResolver(const Maps &maps, const uint depth, const Range_Ptr &source_range, uint &global_minimum_location);
Ranges resolveRange(const Map &map, const Range_Ptr &source_range);
void sortMaps(Maps &maps);

int main()
{
    Ranges seeds;
    Maps maps(7);
    uint global_minimum_location = -1; // not pretty but should work

    readInputFile("input.txt", seeds, maps);
    // sortMaps(maps);

    for (auto &seed : seeds)
    {
        depthFirstResolver(maps, 0, seed, global_minimum_location);
    }
    std::cout << "global minimum: " << global_minimum_location << std::endl;

    return 0;
}

void readInputFile(const std::string &file_location, Ranges &seeds, Maps &maps)
{
    uint map_category_index = 0;
    std::vector<std::string> map_categories = {
        "seed-to-soil map:",
        "soil-to-fertilizer map:",
        "fertilizer-to-water map:",
        "water-to-light map:",
        "light-to-temperature map:",
        "temperature-to-humidity map:",
        "humidity-to-location map:"};

    std::ifstream file(file_location);
    if (file.is_open())
    {
        std::string line;
        if (std::getline(file, line))
        {
            // seeds
            line.erase(0, 7); // cut down 'seeds: '
            std::istringstream iss(line);
            std::vector<std::string> tokens;
            std::string token;
            while (std::getline(iss, token, ' '))
            {
                tokens.push_back(token);
            }
            for (uint i = 0; i < tokens.size(); i = i + 2)
            {
                seeds.push_back(std::make_shared<Range>(std::stoul(tokens[i]), std::stoul(tokens[i + 1])));
            }

            // maps
            while (std::getline(file, line))
            {
                if (line == "")
                {
                    std::getline(file, line);
                    auto it = std::find(map_categories.begin(), map_categories.end(), line);
                    if (it != map_categories.end())
                    {
                        map_category_index = std::distance(map_categories.begin(), it);
                    }
                }
                else
                {
                    std::istringstream iss(line);
                    uint d, s, l;
                    iss >> d >> s >> l;
                    maps[map_category_index].push_back(std::make_shared<MapRange>(d, s, l));
                }
            }
            file.close();
        }
    }
}

void sortMaps(Maps &maps)
{
    for (auto &map : maps)
    {
        std::sort(map.begin(), map.end(), [](const MapRange_Ptr &a, const MapRange_Ptr &b)
                  { return a->source_start < b->source_start; });
    }
}

void depthFirstResolver(const Maps &maps, const uint depth, const Range_Ptr &source_range, uint &global_minimum_location)
{
    auto resolved_ranges = resolveRange(maps[depth], source_range);
    if (depth == 6u)
    {
        for (auto &location_range : resolved_ranges)
        {
            if (location_range->start < global_minimum_location)
            {
                global_minimum_location = location_range->start;
            }
        }
    }
    else
    {
        const uint next_depth = depth + 1u;
        for (auto &resolved_range : resolved_ranges)
        {
            depthFirstResolver(maps, next_depth, resolved_range, global_minimum_location);
        }
    }
}

Ranges resolveRange(const Map &map, const Range_Ptr &source_range)
{
    Ranges resolved_ranges;

    for (const auto &row : map) // use that 'row' ordered
    {
        if (row->source_start <= source_range->start and source_range->start < (row->source_start + row->length))
        {
            const int offset = source_range->start - row->source_start;
            if ((source_range->start + source_range->length) < (row->source_start + row->length))
            {
                resolved_ranges.push_back(std::make_shared<Range>(row->destination_start + offset, source_range->length));
            }
            else
            {
                resolved_ranges.push_back(std::make_shared<Range>(row->destination_start + offset, row->length - offset));
            }
        }
        else if (source_range->start < row->source_start and row->source_start < (source_range->start + source_range->length))
        {
            const int offset = row->source_start - source_range->start;
            if ((source_range->start + source_range->length) < (row->source_start + row->length))
            {
                resolved_ranges.push_back(std::make_shared<Range>(row->destination_start, source_range->length - offset));
            }
            else
            {
                resolved_ranges.push_back(std::make_shared<Range>(row->destination_start, row->length));
            }
        }
    }

    // sort
    // std::sort(resolved_ranges.begin(), resolved_ranges.end(), [](const Range_Ptr &a, const Range_Ptr &b)
    //           { return a->start < b->start; });
    
    // look for unmapped ranges
    if (resolved_ranges.empty())
    {
        resolved_ranges.push_back(source_range);
    }

    return resolved_ranges;
}
